xwiki-platform-core/xwiki-platform-image/xwiki-platform-image-ui/src/main/resources/XWiki/Lightbox/WebHome.xml

```diff
@@ -0,0 +1,514 @@
+        aria-keyshortcuts="F11"
+      &gt;$services.icon.renderHTML('arrows')&lt;/a&gt;
+      &lt;a
+        class="btn download" id="lightboxDownload" aria-controls="blueimp-gallery" target="_blank"
+        aria-label="$escapetool.xml($services.localization.render('image.lightbox.download.title'))"
+        title="$escapetool.xml($services.localization.render('image.lightbox.download.title'))"
+      &gt;$services.icon.renderHTML('download')&lt;/a&gt;
+      &lt;a
+        class="btn autoPlay" aria-controls="blueimp-gallery"
+        aria-label="$escapetool.xml($services.localization.render('image.lightbox.autoPlay.title'))"
+        title="$escapetool.xml($services.localization.render('image.lightbox.autoPlay.title'))"
+        aria-keyshortcuts="Space" aria-pressed="false"
+      &gt;$services.icon.renderHTML('play')&lt;/a&gt;
+      &lt;a
+        class="btn escape" aria-controls="blueimp-gallery"
+        aria-label="$escapetool.xml($services.localization.render('image.lightbox.escape.title'))"
+        title="$escapetool.xml($services.localization.render('image.lightbox.escape.title'))"
+        aria-keyshortcuts="Escape"
+      &gt;$services.icon.renderHTML('cross')&lt;/a&gt;
+    &lt;/div&gt;
+    &lt;div class="lightboxDescription"&gt;
+      &lt;div class="descriptionRow"&gt;
+        &lt;span class="caption"
+          aria-label="$escapetool.xml($services.localization.render('image.lightbox.caption.label'))"&gt;&lt;/span&gt;
+      &lt;/div&gt;
+      &lt;div class="descriptionRow"&gt;
+        &lt;div class="descriptionColLeft"&gt;
+          &lt;span class="title"
+            aria-label="$escapetool.xml($services.localization.render('image.lightbox.fileName.label'))"&gt;&lt;/span&gt;
+        &lt;/div&gt;
+        &lt;div class="descriptionColRight"&gt;
+          &lt;span class="publisher"
+            aria-label="$escapetool.xml($services.localization.render('image.lightbox.publisher.label'))"&gt;&lt;/span&gt;
+          &lt;span class="date"
+            aria-label="$escapetool.xml($services.localization.render('image.lightbox.date.label'))"&gt;&lt;/span&gt;
+        &lt;/div&gt;
+      &lt;/div&gt;
+    &lt;/div&gt;
+    &lt;ol class="indicator"&gt;&lt;/ol&gt;
+  &lt;/div&gt;
+  &lt;!-- The image toolbar popover template. --&gt;
+  &lt;div id="toolbarTemplate"&gt;
+    &lt;div class="btn-group imageToolbar" hidden&gt;
+      &lt;a id="openLightbox"
+        title="$escapetool.xml($services.localization.render('image.lightbox.gallery.title'))"&gt;
+        $services.icon.renderHTML("arrows")
+      &lt;/a&gt;
+      &lt;a class="imageDownload" target="_blank"
+        title="$escapetool.xml($services.localization.render('image.lightbox.download.title'))"&gt;
+        $services.icon.renderHTML("download")
+      &lt;/a&gt;
+    &lt;/div&gt;
+  &lt;/div&gt;
+  {{/html}}
+#end
+{{/velocity}}</content>
+    </property>
+    <property>
+      <extensionPointId>org.xwiki.platform.template.header.after</extensionPointId>
```

Comment:
Why is this after the header of the page? Wouldn't it be better to include this somewhere after the page? I could imagine that otherwise this is potentially bad at least for simple search engines that consider content towards the top as important.

Reply:
> Why is this after the header of the page?

The place where the hidden lightbox template is injected in the HTML doesn't matter that much. What's important is that the JavaScript code is able to find it somewhere in the DOM. The [After Header](https://www.xwiki.org/xwiki/bin/view/Documentation/DevGuide/ExtensionPoint/AfterHeaderUIX) UIXP is used simply because it exists already and is present in all pages for all actions (view, edit, admin, etc.). This UIXP is currently used by many modules to pass hidden data (HTML templates, configuration) from the server-side to JavaScript code.

> Wouldn't it be better to include this somewhere after the page?

Why? There's no "After Page" UIXP currently but we can add it for sure if it brings value.

> I could imagine that otherwise this is potentially bad at least for simple search engines that consider content towards the top as important.

Can you give us a link to a web standard or some search engine documentation that advises against putting hidden HTML templates and configuration inside the BODY tag, close to the top? Or maybe you know some web developer tools that give "search engine" scores that we could use as a metric to find the best location for such hidden data?

Following your logic we should also move stuff like the hidden drawer content after the page since it's definitely not as important as the wiki page content. Same for the top menu bar (as we have on xwiki.org), which can have lots of initially hidden entries that are not as important as the page content. We could also argue that the breadcrumb with its hidden navigation trees is less important than the wiki page content and thus should appear after it in the DOM.

Reply:
The top navigation is wrapped in a `<nav>`-element and thus clearly marked as navigation so I do not think any proper parser will consider this as the main text content. According to its documentation, Google [considers](https://developers.google.com/search/docs/advanced/guidelines/webmaster-guidelines#help-google-understand-your-pages) content hidden inside navigational elements or expanding sections. Google also definitely evaluates CSS and thus knows that the content is hidden so I do not think this is a problem. I was more worried about simpler crawlers that do not consider CSS, e.g., for extracting a description out of the content that they would take this hidden content into account instead of ignoring it. On the other hand, there is a clearly marked `class="content"` which might trigger some heuristics. I'm just considering it good practice to have the main content as early as possible in the document, though definitely not before the navigation which is important, too. Just consider somebody who is using a text browser like lynx for whatever reasons - the navigation is important and okay to come first, but then there should be the main content and not some strange structure you cannot make sense of in a text browser. Note that testing with lynx is something explicitly [recommended by Google](https://developers.google.com/search/docs/advanced/javascript/javascript-seo-basics?hl=en#accessibility). Here two screenshots of the first two pages of https://xwiki.org in Lynx:

![xwiki-org-lynx-1](https://user-images.githubusercontent.com/198317/151342472-f29e1e79-0c46-471f-99f6-d48f073594ca.png)
![xwiki-org-lynx-2](https://user-images.githubusercontent.com/198317/151342477-c66a5130-1c59-4aa8-9ae9-5a4356f36a07.png)

As you can see, there are export dialogs directly visible in the content even though these dialogs are normally only shown when clicking on the export option. While it is okay to have the navigation options visible, I do not think these dialogs should be visible. The same then also applies to any other dialogs like the lightbox. Having them at the end of the page is not that nice either, but imho much better than before the main content.

I think in general we should also consider using the [`<template>`](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/template)-tag whenever possible (I do not expect this is possible here) as this clearly marks the content as template that can be used later by JavaScript. It is not supported in IE, but this shouldn't be an issue anymore.

Reply:
> According to its documentation, Google considers content hidden inside navigational elements or expanding sections.

Yes, but I don't see any explicit advice to put the CSS-hidden content at the end of the page. They only say:

> Make your site's important content visible by default. Google is able to crawl HTML content hidden inside navigational elements such as tabs or expanding sections. However, we consider this content less accessible to users, and recommend that you make your most important information visible in the default page view.

> Note that testing with lynx is something explicitly recommended by Google.

Yes, but besides the "text embedded in images" (which is pretty obvious) there's no real practical advice in:

> One of the easiest ways to test your site's accessibility is to preview it in your browser with JavaScript turned off, or to view it in a text-only browser such as Lynx. Viewing a site as text-only can also help you identify other content which may be hard for Googlebot to see, such as text embedded in images.

What should you look for when you open your web page with Lynx? What makes it accessible and what not. What are the best practices?

Seeing the text from the hidden export modal before the page content with Lynx is definitely not nice but is there a solution besides moving that modal at the end of the page (which obviously doesn't hide it)? I'm fine to start using the ``<template>`` tag in XWiki but Lynx doesn't hide its content so I can only assume that simpler crawlers have the same issue. So the ``<template>`` tag is not a solution currently, but may be in the future.

I think we need to classify the problem:
* is this a search ranking problem? I doubt it, but if it is then we need to decide which search engines do we take into account. Or at least follow some common best practices regarding how web pages are indexed by search engines.
* is this an accessibility problem? If it is then we need to see why our [automated tests](https://github.com/xwiki/xwiki-platform/tree/master/xwiki-platform-distribution/xwiki-platform-distribution-flavor/xwiki-platform-distribution-flavor-test/xwiki-platform-distribution-flavor-test-webstandards) didn't catch it (for the export modal). Moreover, do you know any screen-reader that has issues reading the XWiki web pages because of the hidden content such as the export modal?
* is this a problem with how XWiki pages are rendered in no-JS or text-only browsers? ATM we don't support this use case as per our [Browser Support Strategy](https://dev.xwiki.org/xwiki/bin/view/Community/SupportStrategy/BrowserSupportStrategy).

Reply:
> Seeing the text from the hidden export modal before the page content with Lynx is definitely not nice but is there a solution besides moving that modal at the end of the page (which obviously doesn't hide it)?

You could put the HTML template into a string in JavaScript so it is only created when JavaScript is enabled.

I think there is a huge difference between supporting all features in a browser and between supporting basic page viewing in a browser. Anyways, we do not need to discuss this here, if using this UIXP is common practice for putting templates this is fine for me even though I don't particularly like it, but that's my personal opinion which is guided by the idea of [progressive enhancement](https://developer.mozilla.org/en-US/docs/Glossary/Progressive_Enhancement).

Reply:
> What should you look for when you open your web page with Lynx? What makes it accessible and what not. What are the best practices?

Google [says](https://developers.google.com/search/docs/advanced/javascript/javascript-seo-basics?hl=en#accessibility): "Viewing a site as text-only can also help you identify other content which may be hard for Googlebot to see, such as text embedded in images." I think an accessible site is one where you can easily read and understand its whole content using a text-only browser like Lynx.

> is this a search ranking problem? I doubt it, but if it is then we need to decide which search engines do we take into account. Or at least follow some common best practices regarding how web pages are indexed by search engines.

I doubt there is a measurable problem, but then again search engine ranking is (for good reasons) not completely public so I do not think this is something we can reliably test. I searched and I couldn't find anything regarding this specific question. However, search engines might take accessibility into account (see next point).

> is this an accessibility problem? If it is then we need to see why our [automated tests](https://github.com/xwiki/xwiki-platform/tree/master/xwiki-platform-distribution/xwiki-platform-distribution-flavor/xwiki-platform-distribution-flavor-test/xwiki-platform-distribution-flavor-test-webstandards) didn't catch it (for the export modal). Moreover, do you know any screen-reader that has issues reading the XWiki web pages because of the hidden content such as the export modal?

I'm not an accessibility expert, but as far as I understand one strategy for blind people to access web pages is to use a braille display with a text browser like lynx. I've just searched a bit more and in fact found that both the [WAI Web Content Accessibility Curriculum](https://www.w3.org/WAI/wcag-curric/sam52-0.htm) and the [Accessibility Guidelines for UN Websites](https://www.un.org/en/webaccessibility/guidelines/footer.html) (click on "‘U’ is for Understandable") explicitly mention that when a document is rendered without style sheets, it must still be possible to read the document. While rendering unimportant hidden content before the main content doesn't prevent users from reading the document, it definitely makes it harder and is very annoying.

> is this a problem with how XWiki pages are rendered in no-JS or text-only browsers? ATM we don't support this use case as per our [Browser Support Strategy](https://dev.xwiki.org/xwiki/bin/view/Community/SupportStrategy/BrowserSupportStrategy).

It might be a good idea to support at least a limited set of features without JavaScript and CSS based on my understanding of the accessibility guidelines. However, as mentioned, I'm not an accessibility expert. Accessibility is a complex topic and from my understanding fully automated tests can only test a part of it.

Reply:
@oanalavinia in order to follow @michitux's suggestion you should:

* put the code that generates the HTML in a Velocity macro
* output instead only a script tag with the JSON configuration for the lightbox widget; see https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-realtime/xwiki-platform-realtime-ui/src/main/resources/XWiki/Realtime/Configuration.xml#L665 for an example
* one configuration item is the HTML template which you get by calling the Velocity macro
* from the JavaScript side you look for the script tag and read the JSON configuration
* read the value of the HTML template configuration and append it to the BODY tag at the end when needed